### PR TITLE
PSM Interop: add support to enable CSM Observability and a new test case

### PR DESCRIPTION
**Pull request by stanley-cheung**
This PR is identical to https://github.com/grpc/grpc/pull/34835, but was created before this repo was forked.

### CODE REVIEW

File: framework/test_app/runners/k8s/k8s_base_runner.py [hunk 1]
```diff
@@ -299,6 +299,34 @@ def _reuse_namespace(self) -> k8s.V1Namespace:
         logger.info("Reusing namespace: %s", self.k8s_namespace.name)
         return self.k8s_namespace.get()
 
+    def _create_pod_monitoring(self, template, **kwargs) -> None:
+        if not kwargs["namespace_name"]:
```
Comment (sergiitk) [inline framework/test_app/runners/k8s/k8s_base_runner.py:303]: Ok, so I double checked and this does not work as intended. 

```py
    303 def _create_pod_monitoring_stk(cls, template, **kwargs) -> None:
--> 304     if not kwargs["namespace_name"]:
    305         raise _RunnerError(
    306             "namespace_name required to create PodMonitoring resource"
    307         )
    308     if not kwargs["deployment_id"]:

KeyError: 'namespace_name'
```

However, I don't think we want to check args in the first place. See my suggestion below.

File: framework/test_app/runners/k8s/k8s_base_runner.py [hunk 2]
```diff
@@ -299,6 +299,34 @@ def _reuse_namespace(self) -> k8s.V1Namespace:
...
+            raise _RunnerError(
+                "deployment_id required to create PodMonitoring resource"
+            )
+        if not kwargs["pod_monitoring_name"]:
+            raise _RunnerError(
+                "pod_monitoring_name required to create PodMonitoring resource"
+            )
+        pod_monitoring = self._create_from_template(
+            template, custom_object=True, **kwargs
+        )
```
Comment (sergiitk) [inline framework/test_app/runners/k8s/k8s_base_runner.py:317]:
```suggestion
    def _create_pod_monitoring(
        self,
        template,
        *,
        namespace_name: str,
        deployment_id: str,
        pod_monitoring_name: str,
        **kwargs,
    ) -> None:
        pod_monitoring = self._create_from_template(
            template,
            custom_object=True,
            namespace_name=namespace_name,
            deployment_id=deployment_id,
            pod_monitoring_name=pod_monitoring_name,
            **kwargs,
        )
```

File: framework/test_app/runners/k8s/k8s_base_runner.py [hunk 3]
```diff
@@ -299,6 +299,34 @@ def _reuse_namespace(self) -> k8s.V1Namespace:
...
+        if pod_monitoring.metadata.namespace != kwargs["namespace_name"]:
+            raise _RunnerError(
+                "PodMonitoring resource created with unexpected namespace: "
+                f"{pod_monitoring.metadata.namespace}"
+            )
```
Comment (sergiitk) [inline framework/test_app/runners/k8s/k8s_base_runner.py:322]: Instead, do a check similar to the one in `_create_gamma_route`; just for `PodMonitoring` kind

```py
        if not (isinstance(route, k8s.GammaHttpRoute) and route.kind == "HTTPRoute"):
            raise _RunnerError(
                f"Expected ResourceInstance[HTTPRoute] to be created from"
                f" manifest {template}"
            )
        if route.metadata.name != kwargs["route_name"]:
            raise _RunnerError(
                "ResourceInstance[HTTPRoute] created with unexpected name: "
                f"{route.metadata.name}"
            )
```

File: framework/test_app/runners/k8s/k8s_xds_client_runner.py [hunk 4]
```diff
@@ -158,8 +159,19 @@ def run(  # pylint: disable=arguments-differ
             config_mesh=config_mesh,
             generate_mesh_id=generate_mesh_id,
             print_response=print_response,
+            enable_csm_observability=enable_csm_observability,
         )
 
+        # Create a PodMonitoring resource if CSM Observability is enabled
+        # This is GMP (Google Managed Prometheus)
+        if enable_csm_observability:
+            self._create_pod_monitoring(
```
Comment (sergiitk) [inline framework/test_app/runners/k8s/k8s_xds_client_runner.py:168]: assign this to the class var, then follow this example:

https://github.com/grpc/psm-interop/blob/ea4f7eb127bdd900042de921cdf2a54321ce63e0/framework/test_app/runners/k8s/gamma_server_runner.py#L256-L260

→

https://github.com/grpc/psm-interop/blob/ea4f7eb127bdd900042de921cdf2a54321ce63e0/framework/test_app/runners/k8s/k8s_base_runner.py#L581-L586

→

https://github.com/grpc/psm-interop/blob/ea4f7eb127bdd900042de921cdf2a54321ce63e0/framework/infrastructure/k8s.py#L610-L620

File: tests/gamma/csm_observability_test.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+logger = logging.getLogger(__name__)
+flags.adopt_module_key_flags(xds_k8s_testcase)
+
+_XdsTestServer = xds_k8s_testcase.XdsTestServer
+_XdsTestClient = xds_k8s_testcase.XdsTestClient
+
+
+class CsmObservabilityTest(xds_gamma_testcase.GammaXdsKubernetesTestCase):
+    def test_ping_pong(self):
```
[the remark this answers is not in the document]
Comment (stanley-cheung) [reply tests/gamma/csm_observability_test.py]: Changed to `test_csm_observability`.

File: framework/test_app/runners/k8s/k8s_xds_client_runner.py [hunk 6]
```diff
@@ -158,8 +159,19 @@ def run(  # pylint: disable=arguments-differ
...
+                "csm/pod-monitoring.yaml",
+                namespace_name=self.k8s_namespace.name,
+                deployment_id=self.deployment_id,
+                pod_monitoring_name="%s-gmp" % self.deployment_id,
```
[the remark this answers is not in the document]
Comment (stanley-cheung) [reply framework/test_app/runners/k8s/k8s_xds_client_runner.py]: Fixed.

File: framework/test_app/runners/k8s/k8s_base_runner.py (hunk 1 above)
Comment (stanley-cheung) [reply framework/test_app/runners/k8s/k8s_base_runner.py]: Fixed

File: framework/test_app/runners/k8s/k8s_base_runner.py (hunk 2 above)
Comment (stanley-cheung) [reply framework/test_app/runners/k8s/k8s_base_runner.py]: Fixed. Thanks for the suggestion starter.

File: framework/test_app/runners/k8s/k8s_base_runner.py (hunk 3 above)
Comment (stanley-cheung) [reply framework/test_app/runners/k8s/k8s_base_runner.py]: Used that pattern in the new commit.

File: framework/test_app/runners/k8s/k8s_xds_client_runner.py (hunk 4 above)
Comment (stanley-cheung) [reply framework/test_app/runners/k8s/k8s_xds_client_runner.py]: Now the cleanup is added. There are a couple more places I need to add but was apparent enough.

File: framework/test_app/runners/k8s/k8s_xds_client_runner.py [hunk 7]
```diff
@@ -206,6 +219,9 @@ def cleanup(self, *, force=False, force_namespace=False):
                 )
                 self._delete_service_account(self.service_account_name)
                 self.service_account = None
+            if self.pod_monitoring:
+                self._delete_pod_monitoring(self.pod_monitoring.metadata.name)
```
Comment (sergiitk) [inline framework/test_app/runners/k8s/k8s_xds_client_runner.py:223]: We should save the generate name to a class field and use it here to take care of a condition where the tests fails (and so starts the cleanup) before the `self.pod_monitoring` is created.

**Review by sergiitk — COMMENTED** [review]
LGTM, let's fix that one comment and I'll run adhoc tests.

File: framework/test_app/runners/k8s/k8s_xds_client_runner.py (hunk 7 above)
Comment (stanley-cheung) [reply framework/test_app/runners/k8s/k8s_xds_client_runner.py]: Done.

File: framework/test_app/runners/k8s/k8s_xds_client_runner.py [hunk 8]
```diff
@@ -206,6 +221,9 @@ def cleanup(self, *, force=False, force_namespace=False):
                 )
                 self._delete_service_account(self.service_account_name)
                 self.service_account = None
+            if self.pod_monitoring:
```
Comment (sergiitk) [inline framework/test_app/runners/k8s/k8s_xds_client_runner.py:224]:
```suggestion
            if self.pod_monitoring or force:
```

File: framework/test_app/runners/k8s/k8s_base_runner.py [hunk 9]
```diff
@@ -299,6 +299,43 @@ def _reuse_namespace(self) -> k8s.V1Namespace:
...
+        ):
+            raise _RunnerError(
+                f"Expected ResourceInstance[PodMonitoring] to be created from"
+                f" manifest {template}"
+            )
+        if pod_monitoring.metadata.namespace != namespace_name:
+            raise _RunnerError(
+                "PodMonitoring resource created with unexpected namespace: "
+                f"{pod_monitoring.metadata.namespace}"
+            )
```
Comment (sergiitk) [inline framework/test_app/runners/k8s/k8s_base_runner.py:331]:
```suggestion
```

No need for this check: something should go horribly wrong for the namespace field to not apply.

**Review by sergiitk — COMMENTED** [review]
- [ ] [grpc/core/v1.60.x/branch/linux/grpc_xds_k8s_lb](https://source.cloud.google.com/results/invocations/eb2dedb7-d7a0-48bf-ac72-22a4cde95f30)
- [ ] [grpc/core/v1.60.x/branch/linux/grpc_xds_k8s_lb](https://source.cloud.google.com/results/invocations/3ab35ca8-2ae9-4368-9966-da9e2931f2d9)
- [ ] [grpc/core/v1.60.x/branch/linux/grpc_xds_k8s_lb](https://source.cloud.google.com/results/invocations/5685ab31-93e1-4838-a9ac-88433fee5365)
- [ ] [grpc/core/v1.60.x/branch/linux/psm-csm](https://source.cloud.google.com/results/invocations/1be90a0e-bcdf-44b6-a18c-462bfff1913e)

**Comment by sergiitk** [comment]
CSM gamma test doesn't look right:

```
I0109 00:06:31.510055 140314271825920 k8s_base_runner.py:723] Deployment psm-grpc-client deleted
I0109 00:06:31.510397 140314271825920 k8s_base_runner.py:677] Deleting PodMonitoring None
I0109 00:06:31.535504 140314271825920 retryers.py:206] Retrying framework.xds_k8s_testcase.IsolatedXdsKubernetesTestCase.cleanup in 10.0 seconds as it raised ValueError: At least one of name|label_selector|field_selector is required.
I0109 00:06:41.545946 140314271825920 k8s_base_runner.py:708] Deleting deployment psm-grpc-client
I0109 00:06:41.570450 140314271825920 k8s_base_runner.py:677] Deleting PodMonitoring None
I0109 00:06:41.570951 140314271825920 retryers.py:206] Retrying framework.xds_k8s_testcase.IsolatedXdsKubernetesTestCase.cleanup in 10.0 seconds as it raised ValueError: At least one of name|label_selector|field_selector is required.
I0109 00:06:51.581279 140314271825920 k8s_base_runner.py:708] Deleting deployment psm-grpc-client
I0109 00:06:51.605189 140314271825920 k8s_base_runner.py:677] Deleting PodMonitoring None
E0109 00:06:51.605738 140314271825920 xds_k8s_testcase.py:709] Got error during teardown
Traceback (most recent call last):
  File "/tmp/tmp.JqYFM47JSX/psm-interop/framework/xds_k8s_testcase.py", line 707, in tearDown
    retryer(self.cleanup)
  File "/tmp/tmp.JqYFM47JSX/psm-interop/venv/lib/python3.10/site-packages/tenacity/__init__.py", line 423, in __call__
    do = self.iter(retry_state=retry_state)
  File "/tmp/tmp.JqYFM47JSX/psm-interop/venv/lib/python3.10/site-packages/tenacity/__init__.py", line 369, in iter
    return self.retry_error_callback(retry_state=retry_state)
  File "/tmp/tmp.JqYFM47JSX/psm-interop/framework/helpers/retryers.py", line 141, in error_handler
    raise RetryError(
framework.helpers.retryers.RetryError: Retry error calling framework.xds_k8s_testcase.IsolatedXdsKubernetesTestCase.cleanup: 3 attempts exhausted. Last exception: ValueError: At least one of name|label_selector|field_selector is required
```

**Comment by sergiitk** [comment]
Fixed the pod monitoring cleanup. Re-running the tests.
- [ ] [grpc/core/v1.60.x/branch/linux/grpc_xds_k8s_lb](https://source.cloud.google.com/results/invocations/4e3adc69-0c99-4dbe-a245-1f2342b003f5)
- [ ] [grpc/core/v1.60.x/branch/linux/grpc_xds_k8s_lb](https://source.cloud.google.com/results/invocations/36a2d4d4-a418-44c2-b8ec-a298c01094e9)
- [ ] [grpc/core/v1.60.x/branch/linux/grpc_xds_k8s_lb](https://source.cloud.google.com/results/invocations/5be5c47e-7680-438b-a8f0-aaa7af8296d3)
- [ ] [grpc/core/v1.60.x/branch/linux/psm-csm](https://source.cloud.google.com/results/invocations/6f986959-8f88-4dc4-b818-7dd239f9a1ca)

**Review by sergiitk — APPROVED** [review]
All good now! Thanks for the contribution!

**Pull request merged**